### ISSUE TITLE
Changing global "class" to "name"

### ISSUE THREAD

**Issue by wendellpiez** [usnistgov/OSCAL#172]
# User Story:

As an OSCAL user, I can recognize OSCAL properties and components not by their "class", but by their "name". So instead of

```
<prop class="number">123</prop>
```

we would have

```
<prop name="number">123</prop>
```

This attribute name change appears warranted in view of improvements to data in catalogs in Sprint 10. Indeed, this attribute has been called `type` and `name` before (in very early versions of OSCAL). Reasons for naming it `class` (largely, HTML alignment) now seem to be unconvincing on balance (in view of improvements and OSCAL's overall development). `name` is also shorter (by a char) and more accurate too. (The only one that is clearly wrong seems to be `type`.)

This attribute does three important things:

* It distinguishes a `prop` element among its siblings (although two sibling properties *may* share the same value) - the same on parts, links, and other elements.
* It serves as a hook or identifier for constraints over the property. For example, `name='number'` might link to a constraint that the value of the `prop` must match regex `^\d+`, or satisfy the XPath expression `. castable as xs:number` (both reasonable tests whether a string can produce a number). These constraints are managed out of line (in the declarations model or by other means).
* It offers hints to the downstream process, developer or user as to the intended semantics of the element (which may indeed be active/effective in certain processing contexts)

The change would be global, so instead of `<part class="guidance">` we would have `<part name="guidance">` etc. etc.

## Goals:

OSCAL data will be succinct and sensible to users; in particular it will avoid noise resulting from seeming to mean things that it does not.

## Dependencies:

This is a simple change to schemas and to the data sets. However, not so simple in tools such as XSLTs and Schematrons, especially the more complex ones, which implement the various hooks on `class`. For things to continue to work, a careful review of the code base is in order, with adjustments made.

## Acceptance Criteria

- [ ] Catalog and profile schemas have been adjusted as needed;
- [ ] Samples, examples and reference data sets have been adjusted and are demonstrated to be valid to the new schemas;
- [ ] XSLTs, Schematrons, and any other current processes have been checked, adjusted and regression tested.

**Comment by wendellpiez** [comment]
@david-waltermire-nist @brianrufgsa @anweiss what would you think of adding `@name` and leaving `@class` in place, so both are available?

So one could have

```
<prop name="foo" class="bar baz">VALUE</prop>
```

There are tradeoffs here including documenting (and maybe enforcing) the difference(s).

For example, a document could follow a rule (enforced by Schematron not schema) that only one property with a given name can appear as a sibling (all siblings must have unique names), whereas `@class` would be looser. Also `@name` must be a single token whereas `@class` can be overloaded. Etc.

This would add considerable (more) flexibility to our escape hatches, at the price of some (nominal) simplicity.

**Comment by brianrufgsa** [comment]
@wendellpiez I like this suggestion for a few reasons. The biggest is that it allows for things like:

`<prop name="critical-control" class="fedramp">yes</prop> `

At risk of opening a can of worms, is @id already allowed for <part> and <prop>?

Not coincidentally, I note HTML5 elements all have @name, @id, and @class, along with some fairly standard usage of these three attributes. I would suggest having the same is fairly analogous here.

**Comment by wendellpiez** [comment]
@brianrufgsa the closeness to HTML is purposeful. Even if HTML has had a history of confusion between **name** and **id**.

And yes, `@id` is allowed on `prop` and `part`.

**Comment by david-waltermire-nist** [comment]
Discussed this with @wendellpiez. We defined the following attributes:
- name: an identifier for a data object within the scope of its parent (containing) item.
- namespace (or ns): (for properties) qualifies that the _name_ is part of a organizationally maintained set of names.
- class: (for part) identifies that a given name is part of a set of semantically similar items. 

This means that properties will have a required _name_ and an optional _ns_, and a part will have a required _name_ and an optional _class_. This means in our current examples that _class_ will be switched to _name_ in most cases.

**Comment by brianrufgsa** [comment]
@david-waltermire-nist and @wendellpiez, as discussed yesterday, I agree with

- part having @id, @name, @ns, and @class 
- prop having @id, @name, and @ns

Per that conversation, our intention and recommended "best practice" is:
- @ns reflects the organization or compliance regime defining the information needs addressed by the prop or part 
- @class provides a logical grouping of related parts
- @name allows for naming the purpose of the prop or part
- @id allows for unique identification of the prop or part

I recommend @class and @id are optional; however, @name and @ns are required.
For example, an organization such as FedRAMP would consistently use @ns="fedramp" for all prop and part elements added (extended) by FedRAMP. Then FedRAMP becomes responsible for ensuring no ambiguity among the names used in @name. 

This allows different organizations to use @name values that overlap with other organizations because uniqueness is achieved as a result of the @name/@ns pair.

**Comment by wendellpiez** [comment]
I would rather have `@ns` optional. When not given, it is undefined or implicit. There is nothing wrong with that; an organization that requires @ns can do so using extra schema validation. Only `@name` is required. (I would also be okay with this optional but I think the case for requiring it is stronger.)

**Comment by oscalbuilder** [comment]
@wendellpiez I am good with ```@ns``` being optional and ```@name``` being required.

**Comment by david-waltermire-nist** [comment]
@wendellpiez I am good with @ns being optional and @name being required.

**Comment by brianrufgsa** [comment]
@david-waltermire-nist and @wendellpiez 
We have to be careful about how we use the word "optional" for @ns.

There is optional within the metaschema syntax, which @wendellpiez is proposing. In that case, when missing it is assumed to be @ns='oscal' (or whatever value we decide represents OSCAL itself.)
This allows our current use of prop/part I am fine with that.

**However**; when an organization uses part/prop to extend OSCAL, they **must** use the @ns to ensure we can de-conflict two different organizations using the same value for @name. 

To recap, @ns is optional from a syntax-enforcement perspective, but it actually only optional for the OSCAL team. It is mandatory when extending OSCAL.

**Comment by wendellpiez** [comment]
@brianrufgsa users and developers who assume that "valid to the schema" means "everything is okay and ready to go" will have to unlearn that.

By "OSCAL-defined values for 'name'" here (I assume) is meant the flavor or application of OSCAL we have developed to represent SP800-53, correct? (Whereas another kind of catalog might have other names?)

This opens to a larger question: how might organizations define and enforce such sets of local rules, and where? (Internal checks? At the boundaries, with tools they publish? Both?) We know we have Schematron and we can envision other means and methods. Our earlier work on declarations suggests how we might deploy a set of tools to enable organizations and projects to define such sets of rules and enforce them, without having to write deep code, only a set of declarations or a configuration.

One reason for managing all this in the next layer up is that there will be co-occurrence constraints to enforce. For example, `prop` elements with ns='aaa' can have one set of names, while those with ns='xyz' can have another set of names.

The requirement to validate values of `name` in relation to `ns` could be a useful demonstration of a validation that can be done most effectively outside the schema - in part, so that organizations can customize for themselves or specify (and enforce) with their partners.

**Comment by iMichaela** [comment]
## 5/23/2019
@wendellpiez is working on it. Estimates to be done in a few days.

**Comment by david-waltermire-nist** [comment]
The updated model will work like this:

## part

- id (optional):  allows for unique identification of the part
- name (required): identifies the purpose of the part. Duplicate names can appear in a given parent context.
- ns (optional): reflects the organization or compliance regime defining the information needs addressed by the part. If omitted, the `name` is considered to be a valid name in the OSCAL namespace. TODO: We need to define valid values for OSCAL `name`.
- class (optional): provides a logical grouping of related parts. TODO: Need examples, maybe use 800-53a parts.

Current class needs to be mapped to `name`.

## prop
- id (optional):  allows for unique identification of the prop
- name: (required): identifies the purpose of the prop. Duplicate names can appear in a given parent context.
- ns (optional): reflects the organization or compliance regime defining the information needs addressed by the prop. If omitted, the `name` is considered to be a valid name in the OSCAL namespace. TODO: We need to define valid values for OSCAL `name`.

Current class needs to be mapped to `name`.

## For control, subcontrol, group, and param

- class (optional): provides a logical grouping of related items. TODO: Need examples, maybe use 800-53a parts.

**Comment by wendellpiez** [comment]
Note PR linked above. This issue can be closed assuming the data looks good.

**Comment by david-waltermire-nist** [comment]
This has been completed.
